src/v2.rs

```diff
@@ -0,0 +1,957 @@
+use deno_graph::ModuleGraph;
+use futures::future::poll_fn;
+use sha2::Digest;
+use sha2::Sha256;
+use tokio::io::AsyncReadExt;
+use url::Url;
+
+use crate::error::ParseError;
+use crate::Module;
+use crate::ModuleInner;
+use crate::ModuleKind;
+
+pub(crate) const ESZIP_V2_MAGIC: &[u8; 8] = b"ESZIP_V2";
+
+#[derive(Debug, PartialEq)]
+#[repr(u8)]
+enum HeaderFrameKind {
+  Module = 0,
+  Redirect = 1,
+}
+
+/// Version 2 of the Eszip format. This format supports streaming sources and
+/// source maps.
+#[derive(Debug, Default)]
+pub struct EszipV2 {
+  modules: Arc<Mutex<HashMap<String, EszipV2Module>>>,
+  pub ordered_modules: Vec<String>,
+}
+
+#[derive(Debug)]
+enum EszipV2Module {
+  Module {
+    kind: ModuleKind,
+    source: EszipV2SourceSlot,
+    source_map: EszipV2SourceSlot,
+  },
+  Redirect {
+    target: String,
+  },
+}
+
+#[derive(Debug)]
+enum EszipV2SourceSlot {
+  Pending {
+    offset: usize,
+    length: usize,
+    wakers: Vec<Waker>,
+  },
+  Ready(Arc<Vec<u8>>),
+}
+
+impl EszipV2SourceSlot {
+  fn bytes(&self) -> &[u8] {
+    match self {
+      EszipV2SourceSlot::Ready(v) => v,
+      _ => panic!("EszipV2SourceSlot::unwrap() called on a pending slot"),
+    }
+  }
+}
+
+impl EszipV2 {
+  /// Parse a EszipV2 from an AsyncRead stream. This function returns once the
+  /// header section of the eszip has been parsed. Once this function returns,
+  /// the data section will not necessarially have been parsed yet. To parse
+  /// the data section, poll/await the future returned in the second tuple slot.
+  pub async fn parse<R: tokio::io::AsyncRead + Unpin>(
+    mut reader: tokio::io::BufReader<R>,
+  ) -> Result<
+    (
+      EszipV2,
+      impl Future<Output = Result<tokio::io::BufReader<R>, ParseError>>,
+    ),
+    ParseError,
+  > {
+    let mut magic = [0u8; 8];
+    reader.read_exact(&mut magic).await?;
+
+    if magic != *ESZIP_V2_MAGIC {
+      return Err(ParseError::InvalidV2);
+    }
+
+    let header_len = reader.read_u32().await? as usize;
+    let mut header_and_hash = vec![0u8; header_len + 32];
+    reader.read_exact(&mut header_and_hash).await?;
+
+    let header_bytes = &header_and_hash[..header_len];
+
+    let mut hasher = Sha256::new();
+    hasher.update(&header_bytes);
+    let actual_hash = hasher.finalize();
+    let expected_hash = &header_and_hash[header_bytes.len()..];
+    if &*actual_hash != expected_hash {
+      return Err(ParseError::InvalidV2HeaderHash);
+    }
+
+    let mut modules: HashMap<String, EszipV2Module> = HashMap::new();
+
+    let mut read = 0;
+
+    // This macro reads n number of bytes from the header section. If the header
+    // section is not long enough, this function will be early exited with an
+    // error.
+    macro_rules! read {
+      ($n:expr, $err:expr) => {{
+        if read + $n > header_len {
+          return Err(ParseError::InvalidV2Header($err));
+        }
+        let start = read;
+        read += $n;
+        &header_bytes[start..read]
```

Comment:
Since this is a macro, it could created a fixed array and copy `&header_bytes[start..read]` into it. This will eliminate the `.try_into().unwrap()` calls below

Reply:
Can't do, because `read!` is invoked with a non const length, for example `read!(specifier_len, "specifier")`. The length of an array needs to be const.